### ISSUE TITLE
[RFC] FlyDSL RMSNorm backend for ROCm

### ISSUE THREAD

**Issue by jhinpan** [Dao-AILab/quack#178]
Quack's ROCm/FlyDSL RMSNorm work is in scope, provided FlyDSL remains optional and existing CUDA/CuTe users do not acquire new dependencies. This direction was accepted in [the maintainer response below](https://github.com/Dao-AILab/quack/issues/178#issuecomment-5040707527).

This issue now tracks the staged upstream roadmap. The original prototype analysis and older stacked PRs remain useful prior art, but follow-up work will be forward-ported onto the runtime and API that upstream accepts rather than merged mechanically.

## Current status

- [ ] [#190 — gfx950 FlyDSL RMSNorm forward baseline](https://github.com/Dao-AILab/quack/pull/190)
  - optional `flydsl>=0.3,<0.4` dependency
  - backend-dispatched `rmsnorm_fwd` with CuTe-compatible signature and output/alias semantics
  - gfx950 f16/bf16/f32 forward, mixed affine dtypes, bias/residual/rstd, per-head parameters, padded rows, and wide-row reloads
  - shared host runtime for device/arch validation, per-device launchers, streams, placeholders, and row-layout gates
  - functional custom op for cold `torch.compile` and graph capture
  - deliberately excludes backward, autotuning, autograd, and a dedicated reproduction framework

The exact pre-upstream review history and contributor commits are archived in [jhinpan/quack-FlyDSL#17](https://github.com/jhinpan/quack-FlyDSL/pull/17).

### 1. ROCm CI and FlyDSL compatibility

- [ ] Restore a focused import-isolation regression so ROCm never loads the incompatible CuTe MLIR runtime.
- [ ] Test the exact minimum and latest supported stable FlyDSL releases on hosted CI for installation, import selection, and static API compatibility.
- [ ] Define the required gfx950 execution signal for later PRs: an upstream AMD runner when available, or an AMD-owned external/manual status with reproducible commands and environment metadata.
- [ ] Keep the default CUDA installation and existing CUDA/CuTe tests unchanged.

### 2. Deterministic backward kernel

- [ ] Forward-port the deterministic two-stage backward design from [the old direct-kernel PR](https://github.com/jhinpan/quack-FlyDSL/pull/12) onto the runtime introduced by #190.
- [ ] Cover `dx` and parameter-gradient partial/reduce stages against an independent FP32 reference, including wide rows and exact repeat determinism.
- [ ] Keep this PR private/direct: no autotuning and no public autograd integration.

### 3. Forward autotuning

- [ ] Add an opt-in, correctness-gated, L2-cold forward tuner; retain the analytical configuration as the default and in every candidate set.
- [ ] Re-audit the older [forward autotune implementation](https://github.com/jhinpan/quack-FlyDSL/pull/11) against the current stable FlyDSL API instead of inheriting its private Autotuner/JIT surface unchanged.
- [ ] Persist enough decision, toolchain, and candidate identity to make selected winners reproducible without adding work to the warm launch path.

The deterministic backward kernel and forward autotuner can proceed independently after #190 and the CI/compatibility foundation.

### 4. Training and autograd API parity

- [ ] Connect the accepted forward backend and deterministic backward kernel through the public ROCm training path.
- [ ] Match the existing CuTe behavior for saved state, residual/prenorm, per-head affine tensors, output/residual dtype overrides, and `weight_offset`.
- [ ] Cover eager and `torch.compile` forward/backward numerics, aliases, fake/meta behavior, streams, and graph capture.

This depends on the deterministic backward kernel; tuning is not required for the first training-capable API.

### 5. Backward autotuning

- [ ] Reuse the accepted tuning harness after both the deterministic backward kernel and forward tuning infrastructure exist.
- [ ] Keep deterministic analytical geometry as the default when tuning is disabled.
- [ ] Search row, persistent-grid, and parameter-reduction geometry with correctness and workspace gates before timing.

### 6. gfx942 support

- [ ] Add architecture-specific configuration and validation after gfx950 forward/backward API contracts are stable.
- [ ] Validate on MI300X/gfx942 against the same FP32 references and public behavior as gfx950.
- [ ] Publish architecture-matched performance evidence without changing existing CUDA/CuTe behavior.

## Dependency order

```mermaid
flowchart LR
    Forward190["#190 gfx950 forward baseline"] --> RocmCI["ROCm CI and compatibility"]
    RocmCI --> Backward["Deterministic backward"]
    RocmCI --> ForwardTune["Forward autotuning"]
    Backward --> Autograd["Training and autograd parity"]
    Backward --> BackwardTune["Backward autotuning"]
    ForwardTune --> BackwardTune
    Autograd --> Gfx942["gfx942 support"]
```

## Merge criteria for each PR

- existing CUDA/CuTe imports, APIs, and tests remain unchanged;
- FlyDSL stays optional and ROCm-only modules fail before loading the wrong runtime;
- every kernel path is numerically checked against an independent FP32 reference, not only shape/smoke tested;
- cache keys contain every baked specialization value and device-bound artifacts are keyed per device;
- runtime row-layout, descriptor-span, alignment, stream, and architecture assumptions are enforced at the host boundary;
- stable FlyDSL APIs are preferred, with any unavoidable experimental dependency explicitly inventoried;
- performance claims use verified-idle, architecture-matched GPUs and correctness-gated measurements;
- PRs stay reviewable and focused, with the older stacked branches used as design references rather than merge bases.

## Background

FlyDSL already provides relevant upstream building blocks: [RMSNorm forward/backward](https://github.com/ROCm/FlyDSL/pull/795), [residual backward](https://github.com/ROCm/FlyDSL/pull/800), and an [optimized staged-reduction path](https://github.com/ROCm/FlyDSL/pull/855). Quack owns the downstream kernel integration and PyTorch-facing behavior, following the same downstream-owned model used by [AITER's FlyDSL kernels](https://github.com/ROCm/aiter/tree/main/aiter/ops/flydsl/kernels).

More detailed FlyDSL-side analysis remains in [ROCm/FlyDSL#749](https://github.com/ROCm/FlyDSL/issues/749).

**Comment by tridao** [comment]
Yes this is in scope. We can gate for the features that FlyDSL impl. As long as existing users don’t have to install extra things (similar to how we support softmax in jax in a separate file so PyTorch users don’t have to install jax)
